rooms CRUD 구현

yxxjxxeee opened the pull request:
## #️⃣ 연관된 이슈
- #6

## 📝 작업 내용
> 방 CRUD 기능 구현

## 💬 리뷰 요구사항 (선택)
> 방 전체 조회의 커서 기반 페이지네이션 로직이 올바르게 구현되었는지 확인 부탁드립니다.

yxxjxxeee committed:
feat: (#6) rooms CRUD 구현

gwgw123 commented on src/rooms/rooms.repository.ts:
```diff
@@ -0,0 +1,63 @@
+  ): Promise<{ data: Room[]; hasNextData: boolean; nextCursor: number | null }> {
+    const { name, mode, status, cursor } = query;
+
+    const where: Prisma.RoomWhereInput = { deletedAt: null };
+    if (name) where.name = { contains: name };
+    if (mode) where.mode = mode;
+    if (status) where.status = status;
```
폴리카소를 게임형 웹으로 만든다고 하셨던거 같은데
전체 방 조회시 status가 wating인 방이 위에 보이도록 하면 어떨까요?

그리고 deletedAt을 두신 이유가 뭔가요?? 용도가 궁금하네요

yxxjxxeee replied:
유저 입장에서 바로 참여 가능한 방이 위에 보이면 훨씬 편리할 것 같네요! 좋은 의견 감사해요 ㅎㅎ
그리고 deletedAt을 둔 이유는 전적 플레이나 랭킹 데이터를 보존하기 위해서입니다.
방을 실제로 삭제하면 관련 기록이 사라질 수 있기 때문에 deletedAt으로 soft delete 처리해
삭제된 방은 목록에서 안 보이게 하고 기록은 안전하게 남겨두려는 의도입니다.

gwgw123 replied:
넵 그렇군여
제가 느끼기엔 room은 일시적으로 사용되는 느낌인데 deletedAt을 붙여 전적 기록용으로 겸해서 사용하려는 의도시네요

장기적으로 보관되어야 하는 전적 기록은 별도의 MatchResult 같은 테이블이 생겨도 좋을것 같습니다
의견이니 참고만 해주세요~

gwgw123 replied:
그리고 UserProfile에 코인이나 경험치 컬럼이 있던데
개인전이나 팀 게임에서 순위 결과에 따라 보상이 차등지급 되는 경우 기록용 테이블이 있으면 더 좋겠네요
세부 기능을 정확히 파악하지 못해서 짐작해서 말씀드립니다

rrgks6221 replied:
저도 @gwgw123 님 말씀처럼 방 자체는 삭제시키도 별도의 기록용 테이블을 두는게 어떤가 싶어요

1. row 수가 늘어남에 따라서 조회 비용이 높아짐
2. 삭제시에 삭제뿐만 아닌 기록용 테이블에 저장해야하는 동작이 추가되지만 조회보다 빈도가 낮음
3. 서비스 특성상(실시간 서비스) 유저에게 즉각적으로 보여지는게 중요하기 때문에 조회 효율이 높아야하고 room을 보관하는 작업(삭제 후 기록 테이블에 저장)같은 경우에는 일반적으로 게임들은 게임 종료 후 로딩이 있기 때문에 사용자에게 나쁜 경험으로 보이긴 어려움

yxxjxxeee replied:
제가 설명을 모호하게 했던 것 같습니다. 전적 기록용 테이블(MatchResult)은 이미 존재하지만 현재 Room과 간접적으로 FK로 연결되어 있어 제약 때문에 Room을 물리적으로 삭제할 수 없었습니다. (MatchResult를 삭제하면 안 되기 때문에)
그래서 삭제 대신 deletedAt으로 soft delete 처리했습니다.
그런데 Match 종료 시 관련 정보를 MatchResult에 복사해 저장하는 구조는 생각하지 못했네요. 반영하겠습니다!

gwgw123 replied:
그렇네요 지금 스키마를 보니 match 테이블과 matchResult 테이블이 있네요
soft delete 방식도 좋아보입니다만 혹시나 물리적 삭제가 안되는게 FK 제약 때문에 그러신거라면 
관계 설정시 onDelete 동작을 설정하시는걸 추천드립니다

예를 들어 room은 임시적 데이터이고, match와 matchResult는 장기적 데이터라고 생각했을때
match 테이블은           ``` room         Room          @relation(fields: [roomId], references: [id], onDelete: SetNull) ```
matchResult 테이블은 ``` match    Match      @relation(fields: [matchId], references: [id], onDelete: Cascade)```
과 같이 room을 물리적 삭제시를 대비해 onDelete를 설정해줄 수 있겠네요
room 물리적 삭제 방식을 선택하신다면 roomMember도 onDelete 설정을 해주시면 좋겠습니다

+ 추가로 userProfile 에 onDelete: Cascade 동작 설정이 필요해보입니다 ~

yxxjxxeee replied:
onDelete에 SetNull이라는 속성이 있군요!
Room을 물리적으로 삭제해도 Match는 남고 roomId만 NULL로 바뀌니까 제가 걱정했던 전적 기록이 날아가는 문제를 해결할 수 있겠네요.
좋은 방법 알려주셔서 감사합니다 ㅎㅎ RoomMember나 UserProfile에도 말씀해주신 대로 적용해보겠습니다!

rrgks6221 commented on package.json:
```diff
@@ -42,7 +46,6 @@
     "eslint-plugin-prettier": "^5.2.2",
     "globals": "^16.0.0",
     "jest": "^30.0.0",
-    "prettier": "^3.4.2",
```
이거 삭제되면 아마 개발자 로컬에 다운로드된 프리티어를 가져와 쓰게될건데 그러면 팀원들끼리 버전이 달라서 포맷팅이 깨질 수 있는 가능성이 있을거같아요

yxxjxxeee replied:
실수로 삭제한 것 같네요! ㅠㅠ 알려주셔서 감사합니다!

rrgks6221 commented on src/common/hashing/password-encoder.service.ts:
```diff
@@ -0,0 +1,14 @@
+import * as bcrypt from 'bcrypt';
+
+@Injectable()
+export class PasswordEncoderService {
+  async hash(password: string): Promise<string> {
+    const saltOrRounds = 10;
```
1. 상수는 UPPER_SNAKE_CASE로 선언하는게 일반적입니다
2. SALT_OR_ROUNDS 를 hash의 지역변수로 둔다면 메서드 실행시마다 변수를 할당하게 되는데 클래스 멤버변수로 두는건 어떨까요?? (성능상 큰 차이는 업음)
   1. hash 메서드에서만 사용하게끔 감추시려는 의도였다면 이부분은 무시하셔도 됩니다

```suggestion
export class PasswordEncoderService {
  private SALT_OR_ROUNDS = 10;

  async hash(password: string): Promise<string> {
```

rrgks6221 commented on src/rooms/dto/create-room.dto.ts:
```diff
@@ -0,0 +1,35 @@
+
+  @IsEnum(RoomMode, { message: '모드는 SOLO 또는 TEAM 중 하나여야 합니다.' })
+  mode: RoomMode;
+
+  @IsInt({ message: '최대 인원은 숫자여야 합니다.' })
+  maxPlayers: number;
```
1명 이상만 허용한다면 Min 데코레이터가 추가되면 좋을거같아요

rrgks6221 commented on src/rooms/dto/create-room.dto.ts:
```diff
@@ -0,0 +1,35 @@
+export class CreateRoomDto {
+  @IsString({ message: '방 이름은 문자열이어야 합니다.' })
+  @IsNotEmpty({ message: '방 이름은 필수입니다.' })
+  @MaxLength(20, { message: '방 이름은 최대 15자까지 입력할 수 있습니다.' })
```
메시지와 벨리데이터 인자가 서로 다른데 맞춰지면 좋을거같습니다

yxxjxxeee replied:
인자 수정하는 걸 깜빡했네요! 다음부턴 더 신경 쓰겠습니다

rrgks6221 commented on src/rooms/dto/create-room.dto.ts:
```diff
@@ -0,0 +1,35 @@
+
+  @IsOptional()
+  @IsString({ message: '비밀번호는 문자열이어야 합니다.' })
+  @Matches(/^\d{4}$/, {
+    message: '비밀번호는 숫자 4자리만 가능합니다.',
+  })
+  password?: string;
```
password는 private room이면 필수, public room이면 옵션인건가요??

yxxjxxeee replied:
private room이면 필수, public room이면 비밀번호를 허용하지 않는데
지금은 DTO에서는 단순히 형식만 체크하고 해당 검증 로직은 서비스 단에서 처리하고 있습니다. 
처음에는 커스텀 밸리데이션을 만들어 DTO에서 처리했었는데, 이후 서비스 단으로 옮겼습니다. 
혹시 이런 경우에는 DTO랑 서비스 중에 어디에서 하는게 더 적절할까요?

kimjiwonpg98 replied:
제 생각에는 서비스에서 비즈니스 로직을 실행하기 때문에 서비스에서 허용 여부를 체크하는게 적절해보입니다

추가적으로 class-validator에 @IsNumberString() 가 있는걸로 아는데 그걸 활용해도 좋아보이네요

yxxjxxeee replied:
service에서 비즈니스 로직으로 체크하는 게 적절하다는 의견에 동의합니다.
다만 Entity를 활용하면 비밀번호 필수 여부 같은 룰도 Entity 안에서 관리할 수 있을 것 같아서 그쪽에서 처리하는 방식도 고려해보려 합니다.

rrgks6221 replied:
지원이 얘기대로 서비스에서 체크하는게 바람직하다고 보이고 빠른 피드백을 위해 DTO에서도 체크한다면 커스텀 벨리데이터를 만들어서 체킹할수도있을거같아요
service: 필수
DTO: 옵션

rrgks6221 commented on src/rooms/dto/create-room.dto.ts:
```diff
@@ -0,0 +1,35 @@
+  MaxLength,
+} from 'class-validator';
+
+export class CreateRoomDto {
```
전체적으로 class-validator에 message를 상세하게 적어주셨는데요
개인적으로 class-validator에서 자동으로 생성해주는 message도 꽤 친절하게 만들어준다고 생각해서 예외적인 상황이 아니라면 기본값으로 두는것도 팀원들과 얘기해보시면 좋을거같아요. (메시지도 신경쓰면 유지보수할게 많아짐)

yxxjxxeee replied:
안 그래도 message를 상세하게 적는 건 좀 과한가 싶었어서 @dlehdhun 님과 상의해보겠습니다!

rrgks6221 commented on src/rooms/rooms.controller.ts:
```diff
@@ -0,0 +1,57 @@
+  async findAll(@Query() query: QueryRoomDto) {
+    const { data, hasNextData, nextCursor } = await this.roomsService.findAll(query);
+
+    return {
+      rooms: ResRoomDto.transformRoomEntitiesToDto(data),
+      hasNextData,
+      nextCursor,
+    };
```
커서 관련 응답 DTO도 생성하여 활용하면 좋을거같아요

rrgks6221 commented on src/rooms/rooms.repository.ts:
```diff
@@ -0,0 +1,63 @@
+import { QueryRoomDto } from './dto/query-room.dto';
+
+@Injectable()
+export class RoomsRepository {
```
custom repository 방식을 사용한다면 repository의 인풋을 dto보다는 따로 인터페이스를 설계하는건 어떤가 싶어요
dto를 repository에서 받는다면 controller당 repository가 결합되고 자연스럽게 중간 레이어인 service 레이어도 결합되버림

yxxjxxeee replied:
이해했습니다! 말씀하신 것처럼 Repository가 DTO에 직접 의존하면 Controller와 결합되고 Service까지 자연스럽게 결합될 수 있겠네요. Repository 전용 인터페이스를 설계해서 Service에서 DTO를 변환 후 Repository로 전달하는 방식으로 구조를 개선해보겠습니다.

electrohyun commented on src/rooms/dto/create-room.dto.ts:
```diff
@@ -0,0 +1,35 @@
+} from 'class-validator';
+
+export class CreateRoomDto {
+  @IsString({ message: '방 이름은 문자열이어야 합니다.' })
+  @IsNotEmpty({ message: '방 이름은 필수입니다.' })
+  @MaxLength(20, { message: '방 이름은 최대 15자까지 입력할 수 있습니다.' })
+  name: string;
+
+  @IsEnum(RoomMode, { message: '모드는 SOLO 또는 TEAM 중 하나여야 합니다.' })
+  mode: RoomMode;
+
+  @IsInt({ message: '최대 인원은 숫자여야 합니다.' })
+  maxPlayers: number;
+
+  @IsOptional()
+  @IsBoolean({ message: '비공개 여부는 true 또는 false이어야 합니다.' })
+  isPrivate?: boolean;
+
+  @IsOptional()
+  @IsString({ message: '비밀번호는 문자열이어야 합니다.' })
+  @Matches(/^\d{4}$/, {
+    message: '비밀번호는 숫자 4자리만 가능합니다.',
+  })
+  password?: string;
+}
```
@yxxjxxeee class validator를 통한 유효성 규칙이 보기 쉽게 작성되어있네요! 우선 작업 정말 고생많으셨습니다. 

그런데 프론트에서는 서버가 응답한 에러를 UI에 반영할 때 메세지 문자열을 하드코딩 해야 하는 문제가 있어요! 저희가 미니때 로그인과 회원가입 도메인 작업했던 것처럼 에러가 발생한 부분을 예를 들어 `field` 배열같은 속성에 포함하여 보내주시면 더 안정적으로 에러를 처리할 수 있을 거 같습니다. 바쁘시겠지만 검토 부탁드립니다!

kimjiwonpg98 replied:
지금처럼 서버 측에서 에러 팝업 메시지를 모두 내려준다던지
서버에서 에러 코드를 내려주면 프런트에서 매핑해서 보여준다던지
이건 규칙정하기 나름이라 논의해서 잘 정해지면 좋을 것 같네요

그리고 ValidatePipe 같은거 하나 해두면 좋을 것 같습니다 
https://docs.nestjs.com/techniques/validation

yxxjxxeee replied:
@electrohyun 말씀하신 대로 지금은 프론트에서 에러 메시지를 하드코딩해야 하는 문제가 있네요. 동훈님과 상의해보도록 하겠습니다!
그리고 @kimjiwonpg98 님 말씀처럼 ValidatePipe 문서 참고해서 일관된 에러 응답 구조를 만들어보겠습니다. 좋은 의견 감사합니다 ㅎㅎ

rrgks6221 commented on src/rooms/rooms.service.ts:
```diff
@@ -0,0 +1,101 @@
+import { RoomMode } from '@prisma/client';
+
+@Injectable()
+export class RoomsService {
```
개인적으로 entity를 활용하면 어떤가 싶어요
응답 dto의 transformRoomEntitiesToDto를 보면 엔티티에 대해 생각은 하신거같은데 prisma 특성상 entity 객체에 대한 상태변화 생성같은 메서드를 생성하기 어렵기 때문에 그대로 엔티티로 사용하기에는 아쉬운 부분이 많은거같습니다.

entity를 사용한다면 room(도메인)에 대한 비즈니스 로직을 응집하기 쉬워짐
물론 당장에는 코드 양도 늘어나고 학습 곡선도 늘어나긴 합니다

아래처럼 엔티티를 활요하면 room에 대한 깨지면 안되는 규칙을 한곳에 모아둘 수 있습니다

```ts
class Room {
  constructor() {
    // 생성자 함수
  }

  static  create() {
    this.validate();
    return new Room()
  }

  validate() {
    validateRoomCapacity();
    // 공개 비공개에 따른 패스워드 설정 정책
  }
}
```

yxxjxxeee replied:
아 이 피드백을 지금 확인했네요! 검증 메서드를 어디에 둘지 고민이 많았는데 말씀해주신 것처럼 Entity를 활용하면 Room 도메인에 대한 비즈니스 로직을 한 곳에 응집할 수 있겠네요.

rrgks6221 replied:
설계나 구조 관련된 피드백은 아무래도 학습 시간, 설계시간, 구현시간들이 필요해서 시간이 오래 걸릴 수 있는데요.
프로젝트에서는 정해진 일정이 있기때문에 일정에 차질이 생긴다면 기능상 문제가없는지 체크와 이슈업정도만 해놓고 추후에 개선해도 됩니다. 나중으로 미룰수록 범위가 커지기는해서 여유가 있다면 빨리할수록 좋긴해요

yxxjxxeee committed:
refactor: (#6) Rooms 모듈 구조 및 Repository 의존성 개선

- PasswordEncoderService 상수 캡슐화 개선
- 커서 기반 조회에 ResCursorDto 적용
- RoomsRepository DTO 의존 제거
- CreateRoomDto 유효성 수정
- Prettier 복원
